### PR TITLE
iOS Image Classifier Implementation Files

### PR DESCRIPTION
**Pull request by priankakariatyml**
Added implementation files.
Modified build files

### CODE REVIEW

File: tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+                      error:error];
+
+  return imageClassifier;
+}
+
++ (nullable instancetype)imageClassifierWithOptions:(nonnull TFLImageClassifierOptions *)options
+                                              error:(NSError **)error {
+  TfLiteImageClassifierOptions cOptions = TfLiteImageClassifierOptionsCreate();
+
+  cOptions.classification_options.score_threshold = options.classificationOptions.scoreThreshold;
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m:122]: To avoid having to repeat this copy logic in all tasks that involve classification, what about implementing a objC->C copy method in the TFLImageClassifierOptions and TFLBaseOptions class? You can put it in an extension (e.g. TFLImageClassifierOptions+C.h) to avoid exposing the API to the ObjC users.

File: tensorflow_lite_support/ios/task/vision/utils/sources/GMLImageUtils.m [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+                              .height = height,
+                              .width = width,
+                              .rowBytes = destinationBytesPerRow};
+
+  vImage_Error convertError = kvImageNoError;
+  convertError = vImageConvert_BGRA8888toRGB888(&srcBuffer, &destBuffer, kvImageNoFlags);
+
+  if (convertError != kvImageNoError) {
+    [TFLCommonUtils customErrorWithCode:TFLSupportErrorCodeImageProcessingError
+                            description:@"Image Format Conversion Failed."
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/vision/utils/sources/GMLImageUtils.m:68]: nit: Let's unify the format of description to a normal sentence case. (e.g. Image format conversion failed.)

File: tensorflow_lite_support/ios/task/vision/utils/sources/GMLImageUtils.m [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    buffer_to_return =
+        [UIImage populateRGBBufferFromSourceRGBABuffer:CGBitmapContextGetData(context)
+                                                 width:width
+                                                height:height];
+    CGContextRelease(context);
+  }
+
+  if (buffer_to_return == NULL) {
+    [TFLCommonUtils customErrorWithCode:TFLSupportErrorCodeImageProcessingError
+                            description:@"Image Format Conversion Error"
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/task/vision/utils/sources/GMLImageUtils.m:120]: ditto

**Review by khanhlvg — CHANGES_REQUESTED** [review]

File: tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+}
+@end
+
+@implementation TFLImageClassifier
+- (void)dealloc {
+  TfLiteImageClassifierDelete(_imageClassifier);
+}
+
++ (TFLImageClassifierOptions *)defaultImageClassifierOptionsWithModelPath:
+    (nonnull NSString *)modelPath {
```
[the remark this answers is not in the document]
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m]: I have made the updates.

File: tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m (hunk 1 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m]: I have added categories for base options and classification options and moved the operations to them.

File: tensorflow_lite_support/ios/task/core/sources/TFLBaseOptions.m [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================*/
+#import "tensorflow_lite_support/ios/task/core/sources/TFLBaseOptions.h"
+
+@implementation TFLCpuSettings
+@synthesize numThreads;
+
+- (id)copyWithZone:(NSZone *)zone {
```
Comment (lu-wang-g) [inline tensorflow_lite_support/ios/task/core/sources/TFLBaseOptions.m:20]: Do we need to copy those settings?

File: tensorflow_lite_support/ios/task/processor/sources/TFLClassificationResult.m [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+@synthesize label;
+@synthesize score;
+@synthesize classIndex;
+@end
+
+@implementation TFLClassifications
+@synthesize headIndex;
+@synthesize categories;
+
+- (instancetype)initWithCategories:(NSArray<TFLCategory *> *)categories headIndex:(int)headIndex {
```
Comment (lu-wang-g) [inline tensorflow_lite_support/ios/task/processor/sources/TFLClassificationResult.m:28]: `initWithCategories` may need to update if there's a new property added to `TFLClassifications`, and then it will break existing clients.

Should we just reply on set/get methods generated through synthesize? Same for TFLClassificationResult.

File: tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+@end
+
+@implementation TFLImageClassifierOptions
+@synthesize baseOptions;
+@synthesize classificationOptions;
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    self.baseOptions = [[TFLBaseOptions alloc] init];
```
Comment (lu-wang-g) [inline tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m:37]: This initialization is better moved to TFLBaseOptions and TFLClassificationOptions, so that it can be shared with other tasks. Same for `c/task/vision/image_classifier.cc`, but you can do it in a separate CL for `c/task/vision/image_classifier.cc`.

File: tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    self.baseOptions.modelFile = [[TFLExternalFile alloc] init];
+
+    self.baseOptions.computeSettings = [[TFLComputeSettings alloc] init];
+    self.baseOptions.computeSettings.cpuSettings = [[TFLCpuSettings alloc] init];
+    self.baseOptions.computeSettings.cpuSettings.numThreads = -1;
+
+    self.classificationOptions = [[TFLClassificationOptions alloc] init];
+    self.classificationOptions.maxResults = 5;
```
Comment (lu-wang-g) [inline tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m:45]: maxResults defaults to -1.

File: tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  if (!imageClassifier) {
+    [TFLCommonUtils errorFromTfLiteSupportError:createClassifierError error:error];
+    TfLiteSupportErrorDelete(createClassifierError);
+    return nil;
+  }
+
+  return [[TFLImageClassifier alloc] initWithImageClassifier:imageClassifier];
+}
+
++ (nullable instancetype)imageClassifierWithModelPath:(nonnull NSString *)modelPath
```
Comment (lu-wang-g) [inline tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m:100]: `imageClassifierWithModelPath` can be removed. The reason we have `FooWithModelPath` for text tasks is a purely historical reason. `imageClassifierWithOptions` is good enough.

File: tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m [hunk 10]
```diff
@@ -37,7 +38,9 @@ - (void)testSuccessfullImageInferenceOnMLImageWithUIImage {
   [TFLImageClassifier imageClassifierWithModelPath:self.modelPath error:nil];
   XCTAssertNotNil(imageClassifier);
   
-  UIImage *image = [UIImage imageNamed:@"burger_crop"];
+  NSString *imageName = [[NSBundle bundleForClass:[self class]] pathForResource:@"burger_crop"
```
Comment (khanhlvg) [inline tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m:41]: nit: 
* This should be `imagePath` rather than `imageName` 
* As you access `[[NSBundle bundleForClass:[self class]] pathForResource` quite often, maybe you can create a util method like pathFromBundle:(NSString *)filename

File: tensorflow_lite_support/ios/sources/TFLCommonUtils.m [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    *error = [NSError
+        errorWithDomain:TFLSupportTaskErrorDomain
+                   code:supportError->code
+               userInfo:@{
+                 NSLocalizedDescriptionKey : [NSString stringWithCString:supportError->message
+                                                                encoding:NSUTF8StringEncoding]
+               }];
+}
+
++ (void *)mallocWithSize:(size_t)memSize error:(NSError **)error {
```
[the remark this answers is not in the document]
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/sources/TFLCommonUtils.m:43]: Nope

File: tensorflow_lite_support/ios/task/core/sources/TFLBaseOptions.m (hunk 5 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/core/sources/TFLBaseOptions.m:20]: Yes. When a class is declared as a copy property. This method needs to be implemented.

File: tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m (hunk 9 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m]: Done

File: tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m (hunk 10 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m]: Encapsulated the image initialisation in a method since all that code is repeated

File: tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m (hunk 7 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/vision/sources/TFLImageClassifier.m:37]: I have moved the initialisation to the respective classes.

File: tensorflow_lite_support/ios/task/processor/sources/TFLClassificationResult.m (hunk 6 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/processor/sources/TFLClassificationResult.m:28]: Do you mean to say not have a custom init and just assign the properties from the class in which it is initialised?

Comment (khanhlvg) [reply tensorflow_lite_support/ios/task/processor/sources/TFLClassificationResult.m:28]: Yes that's correct.

File: tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+  XCTAssertNotNil(imageClassifier);
+  GMLImage *gmlImage = [self imageFromBundleWithName:@"burger_crop" ofType:@"jpg"];
+
+  TFLClassificationResult *classificationResults = [imageClassifier classifyWithGMLImage:gmlImage
+                                                                                   error:nil];
+  XCTAssertTrue([classificationResults.classifications count] > 0);
+  XCTAssertTrue([classificationResults.classifications[0].categories count] > 0);
+
+  TFLCategory *category = classificationResults.classifications[0].categories[0];
+  XCTAssertTrue([category.label isEqual:@"cheeseburger"]);
```
Comment (lu-wang-g) [inline tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m:66]: Can we also verify `score` in category?

File: tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+  TFLImageClassifier *imageClassifier =
+      [TFLImageClassifier imageClassifierWithOptions:imageClassifierOptions error:nil];
+  XCTAssertNotNil(imageClassifier);
+
+  GMLImage *gmlImage = [self imageFromBundleWithName:@"burger_crop" ofType:@"jpg"];
+
+  TFLClassificationResult *classificationResults = [imageClassifier classifyWithGMLImage:gmlImage
+                                                                                   error:nil];
+  XCTAssertTrue([classificationResults.classifications count] > 0);
+  XCTAssertTrue([classificationResults.classifications[0].categories count] > 0);
```
Comment (lu-wang-g) [inline tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m:84]: Can we check `count` is maxResults (3)?

File: tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+  CGRect roi = CGRectMake(20, 20, 200, 200);
+  TFLClassificationResult *classificationResults = [imageClassifier classifyWithGMLImage:gmlImage
+                                                                        regionOfInterest:roi
+                                                                                   error:nil];
+  XCTAssertTrue([classificationResults.classifications count] > 0);
+  XCTAssertTrue([classificationResults.classifications[0].categories count] > 0);
+
+  TFLCategory *category = classificationResults.classifications[0].categories[0];
+  XCTAssertTrue([category.label isEqual:@"cheeseburger"]);
```
Comment (lu-wang-g) [inline tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m:110]: Same here and the test below.  Test `score` please.

File: tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m (hunk 13 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m]: Fixed

File: tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m (hunk 14 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m:112]: Fixed

File: tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m (hunk 12 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/test/task/vision/image_classifier/TFLImageClassifierTests.m:66]: Fixed

File: tensorflow_lite_support/ios/task/processor/sources/TFLClassificationResult.m (hunk 6 above)
Comment (priankakariatyml) [reply tensorflow_lite_support/ios/task/processor/sources/TFLClassificationResult.m]: Allright. I have fixed it.

**Review by khanhlvg — APPROVED** [review]
